Hom-ring, ℤ, ℚ, etc. (#1452)
UniMath/agda-unimath: src/linear-algebra/rational-modules.lagda.md

module linear-algebra.rational-modules where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.positive-integers
open import elementary-number-theory.ring-of-rational-numbers

open import foundation.dependent-pair-types
open import foundation.equivalences
open import foundation.function-types
open import foundation.functoriality-dependent-pair-types
open import foundation.logical-equivalences
open import foundation.propositions
open import foundation.subtypes
open import foundation.transport-along-identifications
open import foundation.universe-levels

open import group-theory.abelian-groups
open import group-theory.endomorphism-rings-abelian-groups
open import group-theory.homomorphisms-abelian-groups
open import group-theory.integer-multiples-of-elements-abelian-groups
open import group-theory.isomorphisms-abelian-groups

open import linear-algebra.left-modules-rings
open import linear-algebra.right-modules-rings

open import ring-theory.homomorphisms-rings
open import ring-theory.invertible-elements-rings
open import ring-theory.opposite-rings
open import ring-theory.rational-rings
open import ring-theory.rings
```

</details>

## Idea

> A {{#concept "rational module" Agda=Rational-Module}} is an

Author: The closest reference I could find is this: https://ncatlab.org/nlab/show/rational+vector+space 
Would that be ok?

> A {{#concept "rational module" Agda=Rational-Module}} is an

Author: We don't really have a `Vector-Space` type yet. Would it be ok to add a note explaining that it's equivalent to the concept of rational vector spaces and refactor this later, when we have a better vector spaces infrastructure?

> A {{#concept "rational module" Agda=Rational-Module}} is an

Author: > Ah, right, this is a rational vector space. Then it should be called that

And when me have vector spaces, we might actually want a separate module `rational-vector-spaces` where the distinctions between a "rational module" and a "rational vector space" is essentially the same as the difference with `ring-ℚ` and `discrete-field-ℚ`. 

I'll try to rephrase the idea section, mentioning "rational vector spaces", but I think "rational module" is ok.

> A {{#concept "rational module" Agda=Rational-Module}} is an

Author: See https://github.com/UniMath/agda-unimath/pull/1452/commits/c756d2dcc25b84734e1455ef1a9a71204202fa51

> A {{#concept "rational module" Agda=Rational-Module}} is an

Reviewer: hmm, I'm a bit unsure about wheter this disambiguation is worthwhile, but if you prefer it then okay

[Abelian group](group-theory.abelian-groups.md) whose
[ring of endomorphisms](group-theory.endomorphism-rings-abelian-groups.md) is

> [rational](ring-theory.rational-rings.md). The type of **rational modules** is

Author: see https://github.com/UniMath/agda-unimath/pull/1452/commits/cfa72664ab4c8d9ccdd9e4cfe3d65076a3899202

            ( initial-hom-Rational-Ring ∘ pair (endomorphism-ring-Ab M)))
```

### A rational module is a right module over the ring of rational numbers

```agda
module _
  {l : Level} (M : Rational-Module l)
  where

  right-module-Rational-Module : right-module-Ring l ring-ℚ
  right-module-Rational-Module =
    ( ab-Rational-Module M ,
      initial-hom-Rational-Ring
        ( op-Rational-Ring (rational-ring-endomorphism-Rational-Module M)))
```

### The type of rational modules is equivalent to the type of right modules over the ring of rational numbers

```agda
module _
  {l : Level}
  where

  abstract
    equiv-right-module-Rational-Module :
      right-module-Ring l ring-ℚ ≃ Rational-Module l
    equiv-right-module-Rational-Module =
      equiv-tot
        ( λ M →
          equiv-iff-is-prop
            ( is-prop-has-rational-hom-Ring (op-Ring (endomorphism-ring-Ab M)))
            ( is-prop-is-rational-module-Ab M)
            ( is-rational-is-rational-op-Ring (endomorphism-ring-Ab M) ∘
              is-rational-has-rational-hom-Ring
                ( op-Ring (endomorphism-ring-Ab M)))
            ( initial-hom-Rational-Ring ∘
              op-Rational-Ring ∘

>               pair (endomorphism-ring-Ab M)))

Author: See https://github.com/UniMath/agda-unimath/pull/1452/commits/c756d2dcc25b84734e1455ef1a9a71204202fa51

>               pair (endomorphism-ring-Ab M)))

Reviewer: Ah, I see, nice

open import foundation.universe-levels

open import group-theory.abelian-groups
open import group-theory.endomorphism-rings-abelian-groups
open import group-theory.homomorphisms-abelian-groups
open import group-theory.integer-multiples-of-elements-abelian-groups
open import group-theory.isomorphisms-abelian-groups

open import linear-algebra.left-modules-rings
open import linear-algebra.right-modules-rings

open import ring-theory.homomorphisms-rings
open import ring-theory.invertible-elements-rings
open import ring-theory.opposite-rings
open import ring-theory.rational-rings
open import ring-theory.rings
```

</details>

## Idea

A {{#concept "rational module" Agda=Rational-Module}} is an
[abelian group](group-theory.abelian-groups.md) for which any of the following
[logically](foundation.logical-equivalences.md) equivalent
[propositions](foundation.propositions.md) holds:

- its [ring of endomorphisms](group-theory.endomorphism-rings-abelian-groups.md)
  is [rational](ring-theory.rational-rings.md), i.e., the
  [initial ring homomorphism](elementary-number-theory.ring-of-integers.md) in
  its ring of endomorphisms [inverts](ring-theory.localizations-rings.md) the

>   [positive integers](elementary-number-theory.positive-integers.md);

Reviewer: It doesn't "invert" them, it "makes them invertible", right?

Author: That's the terminology used in the [localizations-rings](https://unimath.github.io/agda-unimath/ring-theory.localizations-rings.html) module but I can use "make them invertible" if you prefer.

Reviewer: Ah, I see, then it's probably fine.

- it is a [left module](linear-algebra.left-modules-rings.md) over the
  [ring of rational numbers](elementary-number-theory.ring-of-rational-numbers.md);
- it is a [right module](linear-algebra.right-modules-rings.md) over the
  [ring of rational numbers](elementary-number-theory.ring-of-rational-numbers.md).

**Note:** Because `ℚ` is a
[discrete field](commutative-algebra.discrete-fields.md), rational modules are
the vector spaces on the
[field of rational numbers](elementary-number-theory.field-of-rational-numbers.md).

## Definitions

> ### The property of being a rational module

Reviewer: ```suggestion
### The predicate on abelian groups of being a rational module
```

```agda
Rational-Module : (l : Level) → UU (lsuc l)
Rational-Module l = type-subtype is-rational-module-prop-Ab

module _
  {l : Level} (M : Rational-Module l)
  where

  ab-Rational-Module : Ab l
  ab-Rational-Module = pr1 M

  is-rational-endomorphism-ring-ab-Rational-Module :
    is-rational-Ring (endomorphism-ring-Ab ab-Rational-Module)
  is-rational-endomorphism-ring-ab-Rational-Module = pr2 M

  rational-ring-endomorphism-Rational-Module : Rational-Ring l
  rational-ring-endomorphism-Rational-Module =
    ( endomorphism-ring-Ab ab-Rational-Module ,
      is-rational-endomorphism-ring-ab-Rational-Module)
```

### The property of being a left module on the rationals

```agda
module _
  {l : Level} (A : Ab l)
  where

  is-rational-left-module-Ab : UU l
  is-rational-left-module-Ab =
    hom-Ring ring-ℚ (endomorphism-ring-Ab A)

  is-prop-is-rational-left-module-Ab :
    is-prop is-rational-left-module-Ab
  is-prop-is-rational-left-module-Ab =
    is-prop-has-rational-hom-Ring (endomorphism-ring-Ab A)

  subtype-is-rational-left-module : Prop l
  subtype-is-rational-left-module =
    ( is-rational-left-module-Ab ,
      is-prop-is-rational-left-module-Ab)
```

> ### The property of being a right module on the rationals

Reviewer: ```suggestion
### The predicate on abelian groups of being a right module on the rationals
```

open import foundation.function-types
open import foundation.functoriality-dependent-pair-types
open import foundation.logical-equivalences
open import foundation.propositions
open import foundation.subtypes
open import foundation.transport-along-identifications
open import foundation.universe-levels

open import group-theory.abelian-groups
open import group-theory.endomorphism-rings-abelian-groups
open import group-theory.homomorphisms-abelian-groups
open import group-theory.integer-multiples-of-elements-abelian-groups
open import group-theory.isomorphisms-abelian-groups

open import linear-algebra.left-modules-rings
open import linear-algebra.right-modules-rings

open import ring-theory.homomorphisms-rings
open import ring-theory.invertible-elements-rings
open import ring-theory.opposite-ring-extensions-rational-numbers
open import ring-theory.opposite-rings
open import ring-theory.ring-extensions-rational-numbers
open import ring-theory.rings
```

</details>

## Idea

A {{#concept "rational module" Agda=Rational-Module}} is an
[abelian group](group-theory.abelian-groups.md) for which any of the following
[logically equivalent](foundation.logical-equivalences.md)

> [propositions](foundation.propositions.md) holds:

Reviewer: ```suggestion
[propositions](foundation.propositions.md) hold:
```

the [initial ring homomorphism](elementary-number-theory.ring-of-integers.md) in
its ring of endomorphisms [inverts](ring-theory.localizations-rings.md) the
[positive integers](elementary-number-theory.positive-integers.md).

This condition is [logically equivalent](foundation.logical-equivalences.md) to
the following [propositions](foundation.propositions.md):

- it is a [left module](linear-algebra.left-modules-rings.md) over the
  [ring of rational numbers](elementary-number-theory.ring-of-rational-numbers.md);
- it is a [right module](linear-algebra.right-modules-rings.md) over the
  [ring of rational numbers](elementary-number-theory.ring-of-rational-numbers.md).

**Note:** Because `ℚ` is a
[discrete field](commutative-algebra.discrete-fields.md), rational modules are
the vector spaces on the
[field of rational numbers](elementary-number-theory.field-of-rational-numbers.md).

## Definitions

### The predicate on abelian groups of being a rational module

```agda
module _
  {l : Level} (A : Ab l)
  where

  is-rational-extension-module-prop-Ab : Prop l

>   is-rational-extension-module-prop-Ab =

Reviewer: yes
